optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java

```diff
@@ -139,42 +145,40 @@ private Constraint transientUsage(ConstraintFactory factory) {
+                        (machineCapacity, processAssignment) -> processAssignment.getUsage(machineCapacity.getResource()))
+                )
+                .filter(((machineCapacity, usage) -> machineCapacity.getSafetyCapacity() < usage))
+                .penalizeLong(MrConstraints.LOAD_COST,
+                              HardSoftLongScore.ofSoft(1L),
+                              (machineCapacity, usage) -> usage - machineCapacity.getSafetyCapacity());
     }
 
     /**
      * availability(r) = capacity(m, r) - usage(m, r)
      * balanceCost = sum(max(0, multiplier * availability(m, r1) - availability(m, r2)))
      */
     private Constraint balanceCost(ConstraintFactory factory) {
-        throw new UnsupportedOperationException("Not yet implemented due to missing support for quad streams.");
-
-        /* TODO: requires quad streams support and groupBy taking two collectors. Alternatively, use a shadow variable.
         return factory.from(MrBalancePenalty.class)
                 .join(MrProcessAssignment.class)
-                .groupBy((penalty) -> penalty, (penalty, assignment) -> assignment.getMachine(),
-                         sumLong((penalty, assignment) -> processAssignment.getUsage(penalty.getOrigin())
-                                 sumLong((penalty, assignment) ->
-                                                 processAssignment.getUsage(penalty.getTarget()))))
-                // QuadStream<MrBalancePenalty, MrMachine, Long, Long>
-                .filter((penalty, machine, originUsage, targetUsage) ->
-                                (machine.getCapacity(penalty.getOrigin()) - originUsage) <
-                                        penalty.getMultiplier() * (machine.getCapacity(penalty.getTarget()) - targetUsage))
-                .penalizeLong(MrConstraintName.BALANCE_COST, HardSoftLongScore.ofSoft(1L), ...);
-         */
+                .groupBy((penalty, processAssignment) -> penalty, (penalty, processAssignment) -> processAssignment.getMachine(),
+                         sumLong((penalty, processAssignment) -> processAssignment.getUsage(penalty.getOriginResource())),
+                         sumLong((penalty, processAssignment) -> processAssignment.getUsage(penalty.getTargetResource()))
+                )
+                .penalizeLong(MrConstraints.BALANCE_COST, HardSoftLongScore.ofSoft(1L), this::balanceCost);
+    }
+
+    private long balanceCost(MrBalancePenalty penalty, MrMachine machine, long originalUsage, long targetUsage) {
+        long originalAvailability =
+                machine.getMachineCapacity(penalty.getOriginResource()).getMaximumCapacity() - originalUsage;
+        long targetAvailability =
+                machine.getMachineCapacity(penalty.getTargetResource()).getMaximumCapacity() - targetUsage;
+        long balanceCost = penalty.getMultiplicand() * originalAvailability - targetAvailability;
+        return balanceCost < 0 ? 0 : balanceCost;
```

Comment:
`Math.max(0, balanceCost)`?

Reply:
Not sure if that's more readable. Or faster.

Reply:
Although performance was not my concern here, although `Math.max()` is intrinsic in Hotspot.
Mainly I'm interested in one less condition we have to test. :-)

Anyway... fine to merge with or without this.

Reply:
Same here, fine to merge with or without this, too. You're owner, your call :)